### DISCUSSION TITLE
How to troubleshoot JavaScript hooks?

### DISCUSSION THREAD

**Discussion by xrd** [pocketbase/pocketbase#3623] [Q&A] [answered]
I'm trying to use a JS hook and finding it difficult to troubleshoot.

First question I have: are hooks supported in the latest version (0.19 as I write this)? As I read the docs and PRs, it seems like there might have been some changes from the original vision (see https://github.com/pocketbase/pocketbase/discussions/2767 and https://github.com/pocketbase/pocketbase/discussions/2825#discussioncomment-6376210). It appears to me that this has been available since 0.17. If something changed between 0.17 and 0.19, I am happy to use whatever is stable as I'm just prototyping right now.

Second, what is the expected output when I run using the `--debug` switch? For example, I am running like this inside my docker container:

```
/pb/pocketbase serve --http=0.0.0.0:8080 --hooksDir /pb/pb_hooks --debug
```

My understanding is that if there is a pb_hooks directory next to the binary then it will use that automatically, but I didn't see a "default" value specified in the help text, so I figured I would be explicit and run it with an absolute path. But, if I left it off, is the correct behavior that pocketbase would use my hooks because that directory is there (even without the switch)?

That invocation prints out a lot more debugging information, but I don't see anything indicating it is loading my JS hooks.

This is the hook I have right now:

```
 # cat /pb/pb_hooks/pb.main.js
onRecordAfterCreateRequest((e) => {
    console.log(e.httpContext)
    console.log(e.record)
    console.log(e.uploadedFiles)
})
```

If I run `node -c pb_hooks/pb.main.js` from outside my container, I get no errors, so I don't think it is a parsing error. But, what would the behavior be if I had a mangled JS file; would pocketbase complain, or silently ignore it?

When I create a new record, I see this, but nothing to indicate it tried or failed to run my hook:

```
[14.00ms] INSERT INTO `hello` (`created`, `hello`, `id`, `updated`) VALUES ('2023-10-27 18:00:02.353Z', 'hello', 'iab9n5a33vucsov', '2023-10-27 18:00:02.353Z')
[0.00ms] SELECT {{_admins}}.* FROM `_admins` WHERE `id`='xe5ky3xnlwwu7ag' LIMIT 1
[0.00ms] SELECT {{_collections}}.* FROM `_collections` WHERE [[id]] = 'bx6gdhz24bbtk00' OR LOWER([[name]])='bx6gdhz24bbtk00' LIMIT 1
[0.00ms] SELECT `hello`.* FROM `hello` ORDER BY [[hello.created]] DESC LIMIT 30
[1.00ms] SELECT COUNT(DISTINCT [[hello.id]]) FROM `hello`

```

What am I looking for to see if hooks are running correctly?

**Comment by ganigeorgiev** [comment, marked as answer]
Only JS hook files that have `*.pb.js` extension will be autoloaded by the "loader" vm.

The easiest way to check whether a hook file was loaded is to add top level `console.log("hello")` and check if the log is printed in the stdout.

I'll consider updating the `--hooksDir` flag helper text, but by default if it is not specified it fallbacks to a relative `pb_data/../pb_hooks` directory. Also note that when specifying flag values you may have to quote them and/or join them with `=` to prevent detecting the value as argument and not as flag value.

**Reply by xrd** [reply]
Doh. Now I see, as the documentation clearly states, that my filename is wrong. Thanks so much.
